### ISSUE TITLE
Cannot use orderby in shortcode, this plugin overrides it

### ISSUE THREAD

**Issue by n9yty** [skyverge/woocommerce-extra-product-sorting-options#17]
I posted here:

https://wordpress.org/support/topic/product-shortcode-in-product-not-respecting-orderby/

But basically, it seems as though this plugin, in the attempt to order everything by the chosen order, is also overriding shortcodes.

In particular we are trying to use the [products] shortcode with a custom orderby, and with this plugin active it is overriding it and instead forcing the default sort to be used regardless of what we specify in the shortcode.

I had suggested a change in that forum post, but I don't know if it would be the right way to handle it or have other side-effects in how the plugin is supposed to operate.

**Comment by bekarice** [comment]
hey @n9yty thanks for the report here! can you please let me know what version of the plugin you're using? Here's what I see using WooCommerce 3.4.5 + WooCommerce Extra Product Sorting Options 2.7.2:

1. Use the **products** shortcode and orderby rating: 
   Input: `[products limit=3 orderby="rating"]`
   Result: I see items sorted properly by rating:
   <img width="844" alt="screen shot 2018-10-02 at 1 38 04 am" src="https://user-images.githubusercontent.com/4983547/46331016-e4b33d00-c5e3-11e8-916f-89bb1e8d5215.png">

1. Input: `[products limit=3 orderby="popularity"]`
   Result: I see items properly sorted by popularity / sales count:
   <img width="831" alt="screen shot 2018-10-02 at 1 38 24 am" src="https://user-images.githubusercontent.com/4983547/46331045-f09eff00-c5e3-11e8-9702-c33973e5f668.png">

1. I go to the Shop page and can confirm my default sorting (menu order) is preserved.

I know you'd noted [this code section](https://github.com/skyverge/woocommerce-extra-product-sorting-options/blob/master/woocommerce-extra-product-sorting-options.php#L347-L360) in the forums, but if you'll look a bit further, you'll see that it also has a safety check for the shortcode arguments to respect them. 

On WooCommerce pages, we have to assume we'll be using the `$_GET` value because these pages will output the sorting dropdown, but on other pages, we should try to use the arguments from the shortcode, and then finally use the default as a fallback.

Could you please clarify what you're seeing here, or where you're using this shortcode so I can try to replicate this issue? thanks!

**Comment by n9yty** [comment]
I also tested various other sort parameters. For example, if I used 'popularity' as you did, limiting it to three, they did sort non-alphabetically. However, if I then put order="DESC" it did not change the sort order.

I didn't try all possible options, but 'date', 'id', 'title' don't work, I don't think we have ratings so that doesn't work. NOthing seems to respect the 'order' ASC/DESC parameter.

**Comment by n9yty** [comment]
Sorry for the noise, but I just read more carefully what you wrote:

"On WooCommerce pages, we have to assume we'll be using the $_GET value because these pages will output the sorting dropdown,"

Our use case is such that we have a WooCommerce product page using a product shortcode, it does not present a sorting drop-down, so $_GET is not present.  The code, as I understand it, in that case just uses the default which is why I had to override it to look for a shortcode defined value.

I'm not saying my patch is the right fix, but I guess you have not expected the use of a [products] shortcode to list other products with a custom sort order on a WooCommerce product page?

**Comment by bekarice** [comment]
@n9yty thanks for the clarification here! I've tested a product page as well, and unfortunately I'm still not able to replicate any of these issues. Here's my results (outputting product details for reference):

1.  `[products ids="53,70,31,56" orderby="menu_order" columns="4" order="ASC"]`

    ✅ Correct
    ✅ Works with order="DESC" as well
    <img width="855" alt="menu-order" src="https://user-images.githubusercontent.com/4983547/46390229-d37e3500-c6a4-11e8-9061-94e9820cca5d.png">

1.  `[products ids="53,70,31,56" orderby="title" columns="4" order="ASC"]`

    ✅ Correct
    ✅ Works with order="DESC" as well
    <img width="840" alt="title" src="https://user-images.githubusercontent.com/4983547/46390260-f872a800-c6a4-11e8-8fd0-e128bc5923e0.png">

1.  `[products ids="53,70,31,56" orderby="date" columns="4" order="ASC"]`

    ✅ Correct
    ✅ Works with order="DESC" as well
    <img width="871" alt="date" src="https://user-images.githubusercontent.com/4983547/46390300-2b1ca080-c6a5-11e8-9f08-7b14d07c792d.png">

1.  `[products ids="53,70,31,56" orderby="rating" columns="4"]`

    ✅ Correct
    ✅ Does not work with DESC because WooCommerce core doesn't accept `order` for `rating` orderby
    <img width="842" alt="rating" src="https://user-images.githubusercontent.com/4983547/46390278-0fb19580-c6a5-11e8-89c9-493eec673a94.png">

1.  `[products ids="53,70,31,56" orderby="popularity" columns="4"]`

    ✅ Correct
    ✅ Does not work with DESC because WooCommerce core doesn't accept `order` for `popularity` orderby
    <img width="843" alt="popularity" src="https://user-images.githubusercontent.com/4983547/46390288-1b9d5780-c6a5-11e8-89a2-956188ff713d.png">

Could you please install the [Transients Manager plugin](https://wordpress.org/plugins/transients-manager/) and use it to delete all transients on your site, ensure all other plugins are deactivated except WooCommerce + Extra Product Sorting Options, then try again? Your site may be returning cached results for these shortcodes, which clearing transients should reveal.

**Comment by n9yty** [comment]
Hi Beka, thanks for digging into this, by the way.

At this point, I don't know what to say, I cannot make it go away and you cannot duplciate it....

I installed Transients Manager and deleted all transients
I switched the theme to the Storefront theme to remove conflicts there
I flushed browser cache, am not running any other caching that I know of
I disabled ALL plugins except WooCommerce and WooCommerce Extra Sorting Options

Problem persists, products still are sorting alphabetically instead of by menu_order with that shortcode.

I deleted the Extra Sorting plugin and re-installed it, no change. Here is the settings column:

<img width="273" alt="image" src="https://user-images.githubusercontent.com/765683/46432236-5d64e700-c713-11e8-99ad-cef33e6d0a5e.png">

If you have any thoughts on how I might help provide more debugging information, let me know. I will keep working on it. Also, if you can help me see the section where you are respecting the parameter, it might help me to debug it a bit more. I'll try below, but if I am missing something please let me know.

As I see it, you are applying add_new_shop_ordering_args() to the filter woocommerce_get_catalog_ordering_args.  This will filter any product request, as I understand it.

In add_new_shop_ordering_args() it looks to see if $_GET['orderby'] is set. I believe this is where you are expecing it should be set for woocommerce product pages, shortcode or otherwise. In our case, looking at headers in the broswer debugger view, this is not set for the request.  The only query string parameters are p, post_type, and preview {as I am looking at a preview of the product, it isn't published). But this is probably for the outer container product, not the shortcode.

So inside add_new_shop_ordering_args() I inserted a debug statement at the top to capture a few things:

```
 $get_contents = print_r( $_GET, true );
 $sort_args_contents = print_r( $sort_args, true );
 echo "DEBUG GET cotents: " . $get_contents;
 echo "DEBUG sort_args cotents: " . $sort_args_contents;
```

In the page output, I see this output twice, once after the page title and once after the product description header, right before the shortcode contents are displayed.

DEBUG GET contents: Array ( [post_type] => product [p] => 1203845 [preview] => true ) 
DEBUG sort_args contents: Array ( [orderby] => menu_order title [order] => ASC [meta_key] => )

If $_GET['orderby'] is not defined, the code after the test would apply the standard sorting.  The next code section that specifically looks for the $sort_args['orderby'] does not apply here becuase is_woocommerce() is true as this is a product page.

So it seems that in your case/testing, the $_GET['orderby'] is set when the shortcode fires, and for some reason it is NOT set here. The question is WHY is it different?

Any thoughts on where to go from here?

**Comment by n9yty** [comment]
@bekarice Maybe you have given up on me since it seems something strange is going on. :)  Is there any more I can do to help debug this?

**Comment by bekarice** [comment]
@n9yty nope, just been out of office for most of the past 2 weeks. :) the additional information here helped! I was able to replicate this after some settings adjustments. Could you apply [this change](https://github.com/skyverge/woocommerce-extra-product-sorting-options/pull/18/commits/85d848a4485c1a58dbfcbb7f4525ffd01e7ca9c6#diff-1f73ee5be748f3090adb4ff70e62ed7bR357) and let me know if it resolves the issue for you? thanks!

**Comment by n9yty** [comment]
Thanks, @bekarice, that takes care of things here for us!  And thanks for your awesome WooSesh presentation.

**Comment by bekarice** [comment]
excellent, we'll get this included in the next release then, thanks for hanging with me here @n9yty  😸 and glad to hear you were able to attend WooSesh! Brian and Patrick did a wonderful job, I appreciate you tuning in! 📺 🍻

**Comment by unfulvio** [comment]
addressed by #18  -- closing this, pending deploy

thanks!

**Comment by EdithAllison** [comment]
Just to say thank you! The official Woo documentation references the incorrect code which results in shortcodes not working, your fix works perfectly, thank you for publishing on github!
